napi/src/threadsafe_function.rs

```diff
@@ -0,0 +1,168 @@
+use crate::{check_status, ptr, sys, Env, Function, Result, Value};
+use std::os::raw::{c_char, c_void};
+
+// TODO Should we pub these enums? or pub the sys mod directly?
+pub use sys::napi_threadsafe_function_call_mode;
+pub use sys::napi_threadsafe_function_release_mode;
+
+pub trait ToJs: Copy + Clone {
+  type Output;
+  type JsValue;
+
+  fn resolve(
+    &self,
+    env: &mut Env,
+    output: &mut Self::Output,
+  ) -> Result<(u64, Value<Self::JsValue>)>;
+}
+
+#[derive(Debug, Clone, Copy)]
+pub struct ThreadsafeFunction<T: ToJs> {
+  raw_value: sys::napi_threadsafe_function,
+  to_js: T,
+}
+
+unsafe impl<T: ToJs> Send for ThreadsafeFunction<T> {}
+unsafe impl<T: ToJs> Sync for ThreadsafeFunction<T> {}
+
+impl<T: ToJs> ThreadsafeFunction<T> {
+  pub fn create(env: Env, func: Value<Function>, to_js: T, max_queue_size: u64) -> Result<Self> {
+    let mut async_resource_name = ptr::null_mut();
+    let s = "napi_rs_threadsafe_function";
+    let status = unsafe {
+      sys::napi_create_string_utf8(
+        env.0,
+        s.as_ptr() as *const c_char,
+        s.len() as u64,
+        &mut async_resource_name,
+      )
+    };
+    check_status(status)?;
+
+    let initial_thread_count: u64 = 1;
+    let mut result = ptr::null_mut();
+    let tsfn = ThreadsafeFunction {
+      to_js,
+      raw_value: result,
+    };
+
+    let ptr = Box::into_raw(Box::from(tsfn)) as *mut _ as *mut c_void;
+
+    let status = unsafe {
+      sys::napi_create_threadsafe_function(
+        env.0,
+        func.raw_value,
+        ptr::null_mut(),
+        async_resource_name,
+        max_queue_size,
+        initial_thread_count,
+        ptr,
+        Some(thread_finalize_cb::<T>),
+        ptr,
+        Some(call_js_cb::<T>),
+        &mut result,
+      )
+    };
+    check_status(status)?;
+
+    Ok(ThreadsafeFunction {
+      to_js,
+      raw_value: result,
+    })
+  }
+
+  pub fn call(
+    &self,
+    value: Result<T::Output>,
+    mode: napi_threadsafe_function_call_mode,
+  ) -> Result<()> {
+    Ok(check_status(unsafe {
+      sys::napi_call_threadsafe_function(
+        self.raw_value,
+        Box::into_raw(Box::from(value)) as *mut _ as *mut c_void,
+        mode,
+      )
+    })?)
+  }
+
+  pub fn acquire(&self) -> Result<()> {
+    Ok(check_status(unsafe {
```

Comment:
just
```rust
check_status(unsafe {
  sys::napi_acquire_threadsafe_function(self.raw_value)
})
```
no need Ok(xxx?)

Reply:
Yes. Will remove all the unnecessary `Ok(...)`.